dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionWriter.java

```diff
@@ -0,0 +1,245 @@
+     * of this method. If it is not removed from the map, then it may be an
+     * obsoleted prediction that needs to be deleted from the database.
+     *
+     * @param prediction new predicted data value.
+     * @param oldPredictionMap existing predicted data values.
+     */
+    private void writePrediction( DataValue prediction, Map<String, DataValue> oldPredictionMap )
+    {
+        boolean predictionIsZeroAndInsignificant = dataValueIsZeroAndInsignificant(
+            prediction.getValue(), prediction.getDataElement() );
+
+        DataValue oldPrediction = oldPredictionMap.get( mapKey( prediction ) );
+
+        if ( oldPrediction == null )
+        {
+            if ( !predictionIsZeroAndInsignificant )
+            {
+                insertPrediction( prediction );
+            }
+        }
+        else
+        {
+            if ( predictionIsZeroAndInsignificant )
+            {
+                if ( !oldPrediction.isDeleted() )
+                {
+                    return; // Keep old prediction in the list to delete it.
+                }
+            }
+            else if ( !prediction.getValue().equals( oldPrediction.getValue() )
+                || oldPrediction.isDeleted() )
+            {
+                updatePrediction( prediction );
+            }
+            else
+            {
+                summary.incrementUnchanged();
+            }
+
+            oldPredictionMap.remove( mapKey( prediction ) );
+        }
+    }
+
+    /**
+     * Adds a predicted data value to the database.
+     * <p>
+     * Note: BatchHandler can be used for inserting only when the period
+     * previously existed. To insert values into new periods (just added to the
+     * database within this transaction), the dataValueService must be used.
+     *
+     * @param prediction the predicted data value.
+     */
+    private void insertPrediction( DataValue prediction )
+    {
+        summary.incrementInserted();
+
+        if ( existingOutputPeriods.contains( prediction.getPeriod() ) )
+        {
+            dataValueBatchHandler.addObject( prediction );
+        }
+        else
+        {
+            dataValueService.addDataValue( prediction );
+        }
+    }
+
+    /**
+     * Updates a predicted data value in the database.
+     *
+     * @param prediction the predicted data value.
+     */
+    private void updatePrediction( DataValue prediction )
+    {
+        summary.incrementUpdated();
+
+        dataValueBatchHandler.updateObject( prediction );
+    }
+
+    /**
+     * (Soft) deletes any remaining old predictions from the database.
+     *
+     * @param oldPredictionMap
+     */
+    private void deleteObsoletePredictions( Map<String, DataValue> oldPredictionMap )
+    {
+        for ( DataValue remainingOldPrediction : oldPredictionMap.values() )
+        {
+            if ( remainingOldPrediction.isDeleted() )
+            {
+                continue;
```

Comment:
Usually java projects try to not use `continue`. I don't know if we follow this trend. In this case for example I'd rather move the 3 lines below into the `if` and negate it.

Reply:
If I look for `continue;` in the project, I see a lot of them. But I've removed this one as you suggested. I try not to let my blocks get too long if I can do an early return/continue, but 3 statements is not too long! Good suggestion.